src/shared/auto-update-cache.ts

```diff
@@ -0,0 +1,58 @@
+import { TimeString } from '@types';
+import ms, { StringValue } from 'ms';
+type Timestamp = number;
+type CacheConstructorParams<Value> = {
+  calculate: () => Promise<Value>;
+  config: AutoUpdateCacheConfig;
+};
+
+const DEFAULT_RETRY_TIME: TimeString = '5M';
+
+export type AutoUpdateCacheConfig = {
+  valid?: 'always' | { onlyFor: StringValue };
+  update: { every: StringValue; ifFailsTryAgainIn?: StringValue };
+};
+
+export class AutoUpdateCache<Value> {
+  private readonly calculate: () => Promise<Value>;
+  private readonly config: AutoUpdateCacheConfig;
+  private cache: { lastUpdated: Timestamp; value: Value | undefined };
+
+  constructor({ calculate, config }: CacheConstructorParams<Value>) {
+    this.calculate = calculate;
+    this.config = config;
+    this.cache = { lastUpdated: 0, value: undefined };
+
+    const isInvalid = config.valid && config.valid !== 'always' && !config.valid.onlyFor;
+
+    if (isInvalid) throw new Error(`'valid' property is misconfigured`);
+
+    this.update();
+  }
+
+  private async update() {
+    const calculated = this.calculate();
+    calculated
```

Comment:
Any particular reason why we are using `then/catch` vs a `try/catch`? Just curious

Reply:
I used the ConcurrentLRUCache as a reference. I think it's pretty similar, right?

Reply:
I think the main difference was that we needed to reference the Promise there. In this case, we just want to wait for it and react to the result/failure